### PR TITLE
Issue 965 - HERE Maps

### PR DESCRIPTION
**Pull request by Charence**
This is for #965 

#### Description
- Endpoints for loading various HERE map tiles (see below)
- Endpoint to list maps available to user
- Frontend changes to use endpoints
- Toggle for multiple map layers
- Access control to Here Maps via hereEnabled flag for users
  - hereEnabled needs to be set to true in customData in system.users for access

#### Test cases

- /T_E_S_T/9197b38e-e1a5-4972-ae25-8be0756151b8/maps/here/16/32735/21789.png - HERE base tile
- /T_E_S_T/9197b38e-e1a5-4972-ae25-8be0756151b8/maps/hereaerial/16/32735/21789.png - HERE satellite tile
- /T_E_S_T/9197b38e-e1a5-4972-ae25-8be0756151b8/maps/heretraffic/16/32735/21789.png - HERE map tile with embedded traffic
- /T_E_S_T/9197b38e-e1a5-4972-ae25-8be0756151b8/maps/heretrafficflow/16/32735/21789.png - HERE traffic flow only tile
- /T_E_S_T/9197b38e-e1a5-4972-ae25-8be0756151b8/maps/herebuildings/51.515279/-0.176694/tile.json - HERE building information for given lat long

#### TODOs
- [ ] Unity needs to be rebuilt prior to merge

### CODE REVIEW

File: backend/middlewares/middlewares.js [hunk 1]
```diff
@@ -181,12 +181,27 @@
 		}
 	}
 
+	function isHereEnabled(req, res, next) {
+		const user = (req.session.user) ? req.session.user.username : "";
+		return User.isHereEnabled(user).then((hereEnabled) => {
```
Comment (carmenfan) [inline backend/middlewares/middlewares.js]: Are you checking the user who is logged to see if they have access to here maps? it should be the teamspace...

i.e If my account isn't here enabled, but TRL account is, I should be able to get HERE maps on TRL models, but not my models within my own account.

File: backend/routes/maps.js [hunk 2]
```diff
@@ -21,15 +21,41 @@ const router = express.Router({mergeParams: true});
...
-	const uri = "/" + req.params.zoomLevel + "/" + req.params.gridx + "/" + req.params.gridy + ".png"
-	systemLogger.logInfo("Fetching osm map tile: " + uri);
+function listMaps(req, res, next) {
+	const user = (req.session.user) ? req.session.user.username : "";
+
+	let maps = [
+		{ name: "Open Street Map", layers: [ { name: "Map Tiles", source: "OSM" } ] }
+	];
+
+	User.isHereEnabled(user).then((hereEnabled) => {
```
Comment (carmenfan) [inline backend/routes/maps.js]: Again, this needs to be the account the model belongs to, not the user who logged in

Comment (jozefdobos) [reply backend/routes/maps.js]: Also for this specific screenshot, the tiles can be much larger. Ideally, the algorithm should somehow try to fill in the most space with a set number of tiles if that makes sense.

**Comment by carmenfan** [comment]
Also can do with some error handling for when here maps isn't in configuration:
image

**Comment by Charence** [comment]
@carmenfan Probably better to disallow HERE if the backend doesn't have config for it; otherwise, the error will be quite unhelpful to the user.

**Comment by carmenfan** [comment]
As you zoom out, the map tiles seems to appear on top of the traffic tiles (may need to adjust the height difference between the 2 depending on zoom level... could be z fighting)
map

**Pull request merged**